### DISCUSSION TITLE
App using Generic Host does not resize

### DISCUSSION THREAD

**Discussion by devployment** [tui-cs/Terminal.Gui#1304] [Q&A]
First of all, I have not much experience with terminal UIs in general. Just checking out this project for an idea I have in mind. 

Currently I'm trying to create a bare bones application following the `ReactiveExample`. As an "app shell" I want to use the .NET generic host (HostBuilder).

https://docs.microsoft.com/en-us/dotnet/core/extensions/generic-host#host-configuration

Problem is that by using the generic host, resizing stops working as you can see here.

https://screencast-o-matic.com/watch/crho2eVfxrm

Maybe thinking to use the generic host is already a bad idea. Not sure. But I wanted to give it a try. If I would have to guess I would say that some kind of message loop does not get the needed information to "resize" in my setup. 

I create it like the following. `AppGui` contains my `Application.Init();` and implements `IHostedService`. The app starts fine and displays the UI. 

Environment 
----
netcoreapp3.1
macOS Big Sur (11.3.1)
iTerm2

Example
----
Program.cs

```
    static class Program
    {
        static void Main(string[] args)
        {
            var cfg = AppConfig.Init();

            CreateHostBuilder(args)
                .Build()
                .Run();
        }

        private static IHostBuilder CreateHostBuilder(string[] args) =>
            Host
                .CreateDefaultBuilder(args)
                .ConfigureServices((hostContext, services) =>
                {
                    services.AddHostedService<AppGui>();
                    services.AddHostedService<Worker>();
                });
    }
```

AppGui.cs

```
    public class AppGui : IHostedService
    {
        private readonly ILogger<AppGui> _logger;

        public AppGui(ILogger<AppGui> logger)
        {
            _logger = logger;
        }

        private static void Start()
        {
            Application.Init();

            RxApp.MainThreadScheduler = TerminalScheduler.Default;
            RxApp.TaskpoolScheduler   = TaskPoolScheduler.Default;

            Application.Run(new AppView(new AppViewModel()));
        }

        public Task StartAsync(CancellationToken cancellationToken)
        {
            Start();
            return Task.CompletedTask;
        }

        public Task StopAsync(CancellationToken cancellationToken)
        {
            _logger.LogInformation("StopAsync");

            return Task.CompletedTask;
        }
    }
```

Maybe somebody has an idea if doing it like this is doable in general.

**Comment by tig** [comment]
I'm not a mac user so can't test myself, but is Terminal.Gui (e.g. with UI Catalog) resizing working well with iTerm2 generally?

I also have no idea what HostBuilder is all about.

**Reply by devployment** [reply]
> but is Terminal.Gui (e.g. with UI Catalog) resizing working well with iTerm2 generally?

Seems to be fine in general. `dotnet run --project UICatalog` works reasonable well I'd say. During resizing the window the content gets scrambled but restores and resizes after stopping the window resizing. Nothing I would be worried about.

> I also have no idea what HostBuilder is all about.

https://docs.microsoft.com/en-us/dotnet/core/extensions/generic-host

```
A host is an object that encapsulates an app's resources, such as:

    Dependency injection (DI)
    Logging
    Configuration
    IHostedService implementations
```

Basically a way to create "worker services" with the setup known from aspnet core. That's basically my idea. Use this worker service approach but with terminal frontend.

But maybe these two ideas just don't mix well.

**Reply by devployment** [reply]
Well, probably https://github.com/migueldeicaza/gui.cs/pull/1278 fixes it. 

I just added `Terminal.Gui.csproj` to my project, and resizing works with my setup by that. 

Correct me if I'm wrong, but there are no "nightly builds" packages available, right?

**Reply by BDisp** [reply]
I thought you were already using the current version of the `main` branch. I really do not know if these changes are being updated on the `NuGet` packages.
Unfortunately that fix is corrupting the clipboard on `Linux`, due to the use of the following:
`Console.Out.Write ("\x1b[?1h")`

**Reply by devployment** [reply]
> I thought you were already using the current version of the main branch

I Was using the published `NuGet` package. But now I am on `main`. 

> Unfortunately that fix is corrupting the clipboard on Linux

As I'm just getting started in general this might be fine for now. Thanks for the hint.
